opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,154 @@
+
+        if self._name_regex.fullmatch(name) is None:
+            raise Exception("Invalid instrument name {}".format(name))
+
+        if len(unit) > 63:
+            raise Exception("unit must be 63 characters or shorter")
+
+        if any(ord(character) > 127 for character in unit):
+            raise Exception("unit must only contain ASCII characters")
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+
+        if not isinstance(callback, Callable):
+            raise Exception("callback must be callable")
+
+        super().__init__(
+            name, unit=unit, description=description, *args, **kwargs
+        )
+
+    def observe(self):
+        return next(self._callback)  # pylint: disable=no-member
+
+
+class Adding(Instrument):
+    pass
+
+
+class Grouping(Instrument):
+    pass
+
+
+class Monotonic(Adding):
+    pass
+
+
+class NonMonotonic(Adding):
+    pass
+
+
+class Counter(Monotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, /, **attributes):
+        if amount < 0:
+            raise Exception("Amount must be non-negative")
+
+        super().add(amount, **attributes)  # pylint: disable=no-member
+
+
+class DefaultCounter(Counter):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def add(self, amount, /, **attributes):
+        return super().add(amount, **attributes)
+
+
+class UpDownCounter(NonMonotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, /, **attributes):
+        pass
+
+
+class DefaultUpDownCounter(UpDownCounter):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def add(self, amount, /, **attributes):
+        return super().add(amount, **attributes)
+
+
+class ObservableCounter(Monotonic, Asynchronous):
+    pass
+
+
+class DefaultObservableCounter(ObservableCounter):
+    def __init__(self, name, callback, unit="", description=""):
+        super().__init__(name, callback, unit=unit, description=description)
+
+
+class ObservableUpDownCounter(NonMonotonic, Asynchronous):
+    pass
+
+
+class DefaultObservableUpDownCounter(ObservableUpDownCounter):
+    def __init__(self, name, callback, unit="", description=""):
+        super().__init__(name, callback, unit=unit, description=description)
+
+
+class Histogram(Grouping, Synchronous):
+    @abstractmethod
+    def record(self, amount, /, **attributes):
+        pass
+
+
+class DefaultHistogram(Histogram):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def record(self, amount, /, **attributes):
+        return super().record(amount, **attributes)
+
+
+class ObservableGauge(Grouping, Asynchronous):
```

Comment:
What does `Grouping` property mean? Why is `Gauge` considered a `Grouping`?

Reply:
It is better explained [here](https://github.com/open-telemetry/opentelemetry-specification/issues/625).